src/lem/eval.rs

```diff
@@ -24,35 +24,62 @@ pub(crate) fn step() -> Result<LEM> {
 mod tests {
     use super::*;
     use crate::field::LurkField;
+    use crate::lem::circuit::SlotsCounter;
     use crate::lem::{pointers::Ptr, store::Store};
     use bellperson::util_cs::{test_cs::TestConstraintSystem, Comparable};
     use blstrs::Scalar as Fr;
 
-    const NUM_INPUTS: usize = 13;
-    const NUM_AUX: usize = 20;
-    const NUM_CONSTRAINTS: usize = 29;
+    const NUM_INPUTS: usize = 1;
+    const NUM_AUX: usize = 19;
+    const NUM_CONSTRAINTS: usize = 17;
+    const NUM_SLOTS: SlotsCounter = SlotsCounter {
+        hash2: 0,
+        hash3: 0,
+        hash4: 0,
+    };
 
     fn test_eval_and_constrain_aux(store: &mut Store<Fr>, pairs: Vec<(Ptr<Fr>, Ptr<Fr>)>) {
         let lem = step().unwrap();
+        lem.check();
+
+        let slots_count = lem.ctl.count_slots();
+
+        assert_eq!(slots_count, NUM_SLOTS);
+
+        let computed_num_constraints = lem.num_constraints::<Fr>(&slots_count);
+
+        // Assures that `MatchSymbol`s will work properly
+        lem.intern_matched_symbols(store);
+
+        let mut all_frames = Vec::default();
+
         for (expr_in, expr_out) in pairs {
-            let witnesses = lem.eval(expr_in, store).unwrap();
+            let frames = lem.eval(expr_in, store).unwrap();
             assert!(
-                witnesses
+                frames
                     .last()
-                    .expect("eval should add at least one step data")
+                    .expect("eval should add at least one frame")
                     .output[0]
                     == expr_out
             );
             store.hydrate_z_cache();
-            for witness in witnesses {
-                let mut cs = TestConstraintSystem::<Fr>::new();
-                lem.constrain(&mut cs, store, &witness).unwrap();
+            let mut cs = TestConstraintSystem::<Fr>::new();
+            for frame in frames.clone() {
+                lem.synthesize(&mut cs, store, &slots_count, &frame)
+                    .unwrap();
                 assert!(cs.is_satisfied());
                 assert_eq!(cs.num_inputs(), NUM_INPUTS);
                 assert_eq!(cs.aux().len(), NUM_AUX);
-                assert_eq!(cs.num_constraints(), NUM_CONSTRAINTS);
+
+                let num_constraints = cs.num_constraints();
+                assert_eq!(computed_num_constraints, num_constraints);
+                assert_eq!(num_constraints, NUM_CONSTRAINTS);
+                // TODO: assert uniformity
             }
+            all_frames.extend(frames);
         }
+
+        lem.assert_all_paths_taken(&all_frames);
```

Comment:
Nit: FYI, not that this matters, because this is a small test:
```
        let mut all_frames = Vec::default();

        for (expr_in, expr_out) in pairs {
            let mut frames = lem.eval(expr_in, store).unwrap();
            assert!(
                frames
                    .last()
                    .expect("eval should add at least one frame")
                    .output[0]
                    == expr_out
            );
            store.hydrate_z_cache();
            let mut cs = TestConstraintSystem::<Fr>::new();
            for frame in frames.iter() {
                lem.synthesize(&mut cs, store, &slots_count, frame).unwrap();
                assert!(cs.is_satisfied());
                assert_eq!(cs.num_inputs(), NUM_INPUTS);
                assert_eq!(cs.aux().len(), NUM_AUX);

                let num_constraints = cs.num_constraints();
                assert_eq!(computed_num_constraints, num_constraints);
                assert_eq!(num_constraints, NUM_CONSTRAINTS);
                // TODO: assert uniformity
            }
            all_frames.extend(frames.drain(..));
        }

```

Also, I wouldn't mind guiding the future implementer on what it would take to "assert uniformity"